src/app/components/weapon-list/weapon-list.component.ts

```diff
@@ -1,52 +1,95 @@
   styleUrls: ['./weapon-list.component.scss'],
 })
-export class WeaponListComponent implements OnInit {
-  listItems?: { [key: string]: any };
-  groupType = '';
+export class WeaponListComponent {
+  readonly filterOptions = FILTER_OPTIONS;
+  readonly sortOptions = SORT_OPTIONS;
+
+  private allData: Array<WeaponSummary> = [];
+  private currentConfig?: FilterSortConfiguration;
+  private currentResults: Array<WeaponSummary> = [];
 
   constructor(
     private impactService: ImpactService,
     public utilityService: UtilityService
-  ) {}
-
-  ngOnInit() {
+  ) {
     this.impactService.getWeapons().subscribe((data: Array<Weapon>) => {
-      this.listItems = new Array();
-
-      this.listItems[WeaponType.Bow] = {
-        mainLabel: 'Bow',
-        weapons: new Array<Character>(),
-      };
-      this.listItems[WeaponType.Catalyst] = {
-        mainLabel: 'Catalyst',
-        weapons: new Array<Character>(),
-      };
-      this.listItems[WeaponType.Claymore] = {
-        mainLabel: 'Claymore',
-        weapons: new Array<Character>(),
-      };
-      this.listItems[WeaponType.Polearm] = {
-        mainLabel: 'Polearm',
-        weapons: new Array<Character>(),
-      };
-      this.listItems[WeaponType.Sword] = {
-        mainLabel: 'Sword',
-        weapons: new Array<Character>(),
-      };
-
-      for (const weaponItem of data) {
-        this.listItems[weaponItem.WeaponType].weapons.push(weaponItem);
-      }
+       this.allData = data.map(item => item.toWeaponSummary());
     });
   }
+
+  get placeholderText() {
+    return "Find a weapon by name…";
+  }
+
+  hasData() {
+    return this.allData.length != 0;
+  }
+
+  updateConfiguration(config: FilterSortConfiguration) {
+    let filteredResults: Array<WeaponSummary> =
+      this.isRefiningSearch(config) ? this.currentResults : this.allData;
+
+    const filterCategory = config.selectedFilter.value;
+    const sort = config.selectedSort.value;
+
+    if (!!filterCategory) {
+      filteredResults = filteredResults.filter(item => item.weaponType === filterCategory);
+    }
+
+    filteredResults = filteredResults.filter(item => item.hasTextInName(config.filterText));
+
+    if (!!sort) {
+      filteredResults = filteredResults.sort(WeaponSummary.getComparator(sort));
+    }
+
+    this.currentConfig = config;
+    this.currentResults = filteredResults;
+  }
+
+  /**
+   * Determines whether the given search configuration is a subset of the one
+   * we have stored locally.
+   */
+  private isRefiningSearch(config: FilterSortConfiguration) {
+    return !!this.currentConfig &&
+      !!config.filterText && config.filterText.includes(this.currentConfig.filterText) &&
+      config.selectedFilter === this.currentConfig.selectedFilter;
+  }
+
+  get results(): Array<WeaponSummary> {
+    return this.currentResults;
+  }
+
+  get resultsMessage() {
+    const plural = this.results.length == 1 ? '' : 's';
+
+    return `${this.results.length} result${plural}`;
+  }
```

Comment:
Probably not worth doing right now, but another option is getting i18n started: https://angular.io/guide/i18n#mark-plurals-and-alternates-for-translation.